In [26]:
# Import the modules
import pandas as pd
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [27]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv("Resources/lending_data.csv")

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [28]:
#check the dataset including the number of loans in the dataset
lending_data_df.count()

loan_size           77536
interest_rate       77536
borrower_income     77536
debt_to_income      77536
num_of_accounts     77536
derogatory_marks    77536
total_debt          77536
loan_status         77536
dtype: int64

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [29]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status'].values.reshape(-1,1)

# Separate the X variable, the features
X = lending_data_df.copy()
X.drop('loan_status', axis=1, inplace=True)

In [30]:
# Review the y variable Series
y[:5]

array([[0],
       [0],
       [0],
       [0],
       [0]], dtype=int64)

In [31]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [32]:
# Split the data using train_test_split
# Assign a random_state of 1 to the function

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [33]:
#check the training dataset including the number of loans
X_train.count()

loan_size           58152
interest_rate       58152
borrower_income     58152
debt_to_income      58152
num_of_accounts     58152
derogatory_marks    58152
total_debt          58152
dtype: int64

In [34]:
#check the testing dataset including the number of loans
X_test.count()

loan_size           19384
interest_rate       19384
borrower_income     19384
debt_to_income      19384
num_of_accounts     19384
derogatory_marks    19384
total_debt          19384
dtype: int64

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [35]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model

classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data

classifier.fit(X_train, y_train.ravel())

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [36]:
# Make a prediction using the testing data

predictions = classifier.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [37]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0","Actual 1"], columns=["Predict 0","Predict 1"]
)

print("Confusion Matrix")
cm_df

Confusion Matrix


Predict 0  Predict 1
Actual 0      18663        102
Actual 1         56        563

In [38]:
# Print the classification report for the model

print("The Classification Report")
print(classification_report(y_test, predictions))

The Classification Report
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** As shown above, the logistic regression model predicts the '0' (healthy loan) with 100% accuracy rate while predict '1'(high-risk loan) with only 88% accuracy rate. Overall, the model can predict both healthy and high-risk loan lables with 99% accuracy rate. 